### PR TITLE
[AMDGPU] Force the third source operand of the MAI instructions to VGPR if no AGPRs are used.

### PR DESCRIPTION
**Pull request by alex-t**
eaf85b9c28 "[AMDGPU] Select VGPR versions of MFMA if possible" prevents the compiler from reserving AGPRs if a kernel has no inline asm explicitly using AGPRs, no calls, and runs at least 2 waves with not more than 256 VGPRs. This, in turn, makes it impossible to allocate AGPR if necessary. As a result, regalloc fails in case we have an MAI instruction that has at least one AGPR operand.
This change checks if we have AGPRs and forces operands to VGPR if we do not have them.

### CODE REVIEW

**Comment by jayfoad** [comment]
> eaf85b9c28 "[AMDGPU] Select VGPR versions of MFMA if possible"

I can't find that commit. Is it upstream?

**Comment by alex-t** [comment]
> I can't find that commit. Is it upstream?
That one is in stg-mainline-open
upstream hash is: aeaf85b9c283c61413c14b52f0f289348eab6a5d

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -14147,9 +14149,14 @@ void SITargetLowering::AdjustInstrPostInstrSelection(MachineInstr &MI,
     // use between vgpr and agpr as agpr tuples tend to be big.
     if (!MI.getDesc().operands().empty()) {
       unsigned Opc = MI.getOpcode();
+      bool NoAGPRs = !Info->mayNeedAGPRs();
+      SmallVector<int> Opnds;
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:14153]: Can you avoid using SmallVector here? I'd add src2 to the static array and break from the loop instead. This is cheaper. Actually I'd probably add isMAI() check too to the whole if (in a separate NFCI).

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -14169,18 +14176,19 @@ void SITargetLowering::AdjustInstrPostInstrSelection(MachineInstr &MI,
...
-      // Resolve the rest of AV operands to AGPRs.
-      if (auto *Src2 = TII->getNamedOperand(MI, AMDGPU::OpName::src2)) {
-        if (Src2->isReg() && Src2->getReg().isVirtual()) {
-          auto *RC = TRI->getRegClassForReg(MRI, Src2->getReg());
-          if (TRI->isVectorSuperClass(RC)) {
-            auto *NewRC = TRI->getEquivalentAGPRClass(RC);
-            MRI.setRegClass(Src2->getReg(), NewRC);
-            if (Src2->isTied())
-              MRI.setRegClass(MI.getOperand(0).getReg(), NewRC);
+      if (!NoAGPRs)
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:14179]: You can do this instead:
```
if (NoAGPRs)
  return;
```

**Review by Toptowie — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp (hunk 1 above)
Comment (alex-t) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp]: > Can you avoid using SmallVector here? I'd add src2 to the static array and break from the loop instead. This is cheaper. Actually I'd probably add isMAI() check too to the whole if (in a separate NFCI).

Done

File: llvm/test/CodeGen/AMDGPU/smfmac_no_agprs.ll [hunk 3]
```diff
@@ -0,0 +1,53 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 2
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx940 -verify-machineinstrs < %s | FileCheck -enable-var-scope -check-prefixes=gfx940 %s
```
Comment (rampitec) [inline llvm/test/CodeGen/AMDGPU/smfmac_no_agprs.ll:2]: The convention for check labels is all-uppercase.

**Review by rampitec — APPROVED** [review]

**Pull request merged**